### ISSUE TITLE
Port Datastore onto GAPIC

### ISSUE THREAD

**Issue by bjwatson** [GoogleCloudPlatform/google-cloud-python#2746]
### What
Port the gRPC code path for Datastore onto GAPIC. The package is available [here](https://pypi.python.org/pypi/gapic-google-cloud-datastore-v1).

### Why
This is not a beta-blocker, since there already is gRPC code for Datastore, but it is preferred.

**Comment by dhermes** [comment]
@bjwatson Two questions:

1. Why doesn't the new package depend on `google-gax>=0.15` (which has the new auth helpers)?
1. Are we in a situation where some of the GAPIC packages depend on `0.13 <= google-gax < 0.14` while others depend on `0.14 <= google-gax < 0.15` (as this one does)?

**Comment by bjwatson** [comment]
@dhermes Only because @geigerj says that he needs to make some small updates to the GAPIC generator for it to be compatible with `google-gax>=0.15`. He might be able to get that done today, and then we can build all packages with the newer `google-gax` dependency.

I thought I would get this in your hands right away, rather than waiting for the necessary fixes to support the newer auth helpers. However, if you think it's a more efficient use of your time to wait, then we can do that.

**Comment by dhermes** [comment]
@bjwatson Sorry I didn't notice this until now, but the [current form][1] of `grpc-google-cloud-datastore-v1` can't run without gRPC. This means we can't get access to the protobufs without forcing `grpc` to be installed.

Right now we support Protobuf-over-HTTP and Protobuf-over-gRPC, so we need access to the protobuf classes in use cases where gRPC isn't available (i.e. App Engine).

The generated protobuf classes currently live in [`datastore._generated`][2], which is how we import them. In order to support gRPC we have `datastore_pb2.py` and `datastore_grpc_pb2.py` in the `_generated` module

/cc @jonparrott 

[1]: https://github.com/googleapis/api-client-staging/blob/f9305a6602771140c954939cf00b8ab09101d0c3/generated/python/grpc-google-cloud-datastore-v1/google/datastore/v1/datastore_pb2.py#L891-L895
[2]: https://github.com/GoogleCloudPlatform/google-cloud-python/tree/a10cf56445b51750c4d9bc1f3b5bde2522f92caf/datastore/google/cloud/datastore/_generated

**Comment by bjwatson** [comment]
@dhermes Does https://github.com/googleapis/api-client-staging/pull/106/commits/6649c45915223fa76fb390f9c12d90cb567a1496 do what you need? Do you also need changes to `setup.py`? I see that your `setup.py` [depends on `grpcio` regardless](https://github.com/GoogleCloudPlatform/google-cloud-python/blob/master/datastore/setup.py#L52-L55).

Note that this will become unnecessary once https://github.com/grpc/grpc/issues/8056 is released in `grpcio==1.1.0`.

**Comment by dhermes** [comment]
Yes it does do what I had in mind.

We **had** conditional checks in our `setup.py`, but @jonparrott discouraged that since it confuses wheel building. But maybe I am wrong on how I remember that.

AFAIK, even though `grpcio` is in our `setup.py` for packages like `google-cloud-datastore`, it doesn't stop from installing into a `lib/` dir in an App Engine app. @jonparrott can confirm or deny or throw his hands up in sadness over packaging?

**Comment by jonparrott** [comment]
`grpcio` will indeed be vendored into `lib` for app engine and explode on import.

**Comment by dhermes** [comment]
@bjwatson Ping me once released?

**Comment by geigerj** [comment]
@bjwatson Is this a Datastore-specific hand-edit, or do we intend to provide these conditional imports across the gRPC packages we maintain? If it's the latter, we should figure out what the timeline for `grpcio 1.1.0` is and whether we need to automate this in our tooling.

**Comment by dhermes** [comment]
@geigerj @bjwatson We would love to see a new release of `grpcio` because of https://github.com/grpc/grpc/issues/7849, so I am on team "pushing for release"

**Comment by bjwatson** [comment]
@dhermes We're going to wait on porting Datastore to GAPIC until after the beta announcement. We discussed and we don't want to take any chances on destabilizing the `google-cloud-python` code for the four APIs. The only one that will go beta with GAPIC is Logging.

**Comment by dhermes** [comment]
SGTM

**Comment by bjwatson** [comment]
@lukesneeringer @dhermes If we're blocking the `x-goog-api-client` header updates on this issue, then it's `P0`.

**Comment by dhermes** [comment]
@bjwatson Those are already in for HTTP. I can get them in to the gRPC transport before the process is finished (#3066 and #3064 will give you an idea of what the process looks like).

**Comment by bjwatson** [comment]
Ok, thanks Danny!

**Comment by lukesneeringer** [comment]
@dhermes This is now done, question mark?

**Comment by dhermes** [comment]
Almost. One follow-up PR, coming in hot.

**Comment by bjwatson** [comment]
Woo hoo! Great job, Danny!

@lukesneeringer @dhermes Does this mean that we'll have the new HTTP headers working for Datastore as soon as this is published?

**Comment by lukesneeringer** [comment]
I believe we have the HTTP headers now. But, yes.
I am thinking a quick umbrella release next week as part of the GA effort, but we can push a 0.23.x for this if we want it sooner.

**Comment by dhermes** [comment]
I just did a

```
$ git diff datastore-0.23.0 -- datastore/google
```

and the only change that is in the **public** interface is the addition of `use_gax` in the `Client` constructor.

**Comment by lukesneeringer** [comment]
Sure, but it is a huge under-the-hood change, because those with GRPC support are suddenly using the GRPC transport now.

**Comment by dhermes** [comment]
@lukesneeringer We already had the gRPC transport (for many releases), it was just the raw stub produced by the `protoc` plugin rather than the `DatastoreClient` which wraps that very same class.

**Comment by lukesneeringer** [comment]
Okay.
